libraries/AP_GPS/AP_GPS.cpp

```diff
@@ -415,6 +420,35 @@ AP_GPS::detect_instance(uint8_t instance)
         goto found_gps;
         break;
 
+#if HAL_WITH_UAVCAN
+    // user has to explicitly set the UAVCAN type, do not use AUTO
+    // do not try to detect the UAVCAN type, assume it's there
+    case GPS_TYPE_UAVCAN:
+        if (AP_BoardConfig::get_can_enable() >= 1) {
+            _broadcast_gps_type("UAVCAN", instance, -1);
+            new_gps = new AP_GPS_UAVCAN(*this, state[instance], nullptr);
+
+            // register new listener at first empty node
+            if (hal.can_mgr != nullptr) {
+                AP_UAVCAN *ap_uavcan = hal.can_mgr->get_UAVCAN();
+                if (ap_uavcan != nullptr) {
+                    ap_uavcan->register_gps_listener(new_gps, 0);
+
+                    if (AP_BoardConfig::get_can_debug() >= 2) {
+                        hal.console->printf("AP_GPS_UAVCAN registered\n\r");
+                    }
+
+                    state[instance].status = NO_GPS;
+                    drivers[instance] = new_gps;
+                    timing[instance].last_message_time_ms = now;
```

Comment:
UAVCAN GPS behavior is very different from other GPS drivers and there is a reason to be different.
However I will take your approach and do changes to make it compliant to old code.
Also in my opinion, the use of goto was very bad decision in first place.